File: backend/src/main/java/com/agentworkbench/session/service/MessageQueueService.java
package com.agentworkbench.session.service;

import com.agentworkbench.session.entity.MessageQueue;
import com.agentworkbench.session.mapper.MessageQueueMapper;
import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.baomidou.mybatisplus.core.conditions.update.LambdaUpdateWrapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class MessageQueueService {

    private final MessageQueueMapper messageQueueMapper;

    /**
     * 入队：将消息加入队列末尾
     */
    public MessageQueue enqueue(Long sessionId, Long userId, String content, String images) {
        // 查询当前队列最大 sortOrder
        MessageQueue last = messageQueueMapper.selectOne(
                new LambdaQueryWrapper<MessageQueue>()
                        .eq(MessageQueue::getSessionId, sessionId)
                        .eq(MessageQueue::getStatus, "PENDING")
                        .orderByDesc(MessageQueue::getSortOrder)
                        .last("LIMIT 1"));
        int maxOrder = last != null ? last.getSortOrder() : 0;

        MessageQueue item = new MessageQueue();
        item.setSessionId(sessionId);
        item.setUserId(userId);
        item.setContent(content);
        item.setImages(images);
        item.setSortOrder(maxOrder + 1);
        item.setStatus("PENDING");
        messageQueueMapper.insert(item);
        return item;
    }

    /**
     * 出队：取出并删除队列头部消息
     */
    public MessageQueue dequeue(Long sessionId) {
        MessageQueue head = messageQueueMapper.selectOne(
                new LambdaQueryWrapper<MessageQueue>()
                        .eq(MessageQueue::getSessionId, sessionId)
                        .eq(MessageQueue::getStatus, "PENDING")
                        .orderByAsc(MessageQueue::getSortOrder)
                        .last("LIMIT 1"));
        if (head != null) {
            head.setStatus("DELETED");
            messageQueueMapper.updateById(head);
        }
        return head;
    }

    /**
     * 删除指定队列消息
     */
    public void delete(Long queueId) {
        MessageQueue item = messageQueueMapper.selectById(queueId);
        if (item != null) {
            item.setStatus("DELETED");
            messageQueueMapper.updateById(item);
        }
    }

    /**
     * 重新排序：将指定消息上移/下移（交换 sort_order）
     */
    public void reorder(Long queueId, String direction) {
        MessageQueue current = messageQueueMapper.selectById(queueId);
        if (current == null || !"PENDING".equals(current.getStatus())) return;

        MessageQueue neighbor;
        if ("up".equals(direction)) {
            neighbor = messageQueueMapper.selectOne(
                    new LambdaQueryWrapper<MessageQueue>()
                            .eq(MessageQueue::getSessionId, current.getSessionId())
                            .eq(MessageQueue::getStatus, "PENDING")
                            .lt(MessageQueue::getSortOrder, current.getSortOrder())
                            .orderByDesc(MessageQueue::getSortOrder)
                            .last("LIMIT 1"));
        } else {
            neighbor = messageQueueMapper.selectOne(
                    new LambdaQueryWrapper<MessageQueue>()
                            .eq(MessageQueue::getSessionId, current.getSessionId())
                            .eq(MessageQueue::getStatus, "PENDING")
                            .gt(MessageQueue::getSortOrder, current.getSortOrder())
                            .orderByAsc(MessageQueue::getSortOrder)
                            .last("LIMIT 1"));
        }

        if (neighbor != null) {
            int tempOrder = current.getSortOrder();
            current.setSortOrder(neighbor.getSortOrder());
            neighbor.setSortOrder(tempOrder);
            messageQueueMapper.updateById(current);
            messageQueueMapper.updateById(neighbor);
        }
    }

    /**
     * 查询队列列表（PENDING 状态，按 sort_order 升序）
     */
    public List<MessageQueue> listPending(Long sessionId) {
        return messageQueueMapper.selectList(
                new LambdaQueryWrapper<MessageQueue>()
                        .eq(MessageQueue::getSessionId, sessionId)
                        .eq(MessageQueue::getStatus, "PENDING")
                        .orderByAsc(MessageQueue::getSortOrder));
    }

    /**
     * 根据 ID 查询
     */
    public MessageQueue getById(Long queueId) {
        return messageQueueMapper.selectById(queueId);
    }

    /**
     * 清空队列
     */
    public void clear(Long sessionId) {
        messageQueueMapper.update(null,
                new LambdaUpdateWrapper<MessageQueue>()
                        .eq(MessageQueue::getSessionId, sessionId)
                        .eq(MessageQueue::getStatus, "PENDING")
                        .set(MessageQueue::getStatus, "DELETED"));
    }
}

File: backend-ts/src/session/message-queue.service.ts
import type { MessageQueue } from './types.js';
import type { MessageQueueRepository } from './message-queue.repository.js';
export class MessageQueueService {
  constructor(private readonly repo: MessageQueueRepository) {}

  async enqueue(sessionId: number, userId: number, content: string, images: string | null): Promise<MessageQueue> {
    // 事务 + FOR UPDATE 锁住队尾，避免并发 enqueue 读到相同 max(sort_order) 产生重复排序值
    return this.repo.transaction(async (tx) => {
      const last = await tx.findLastPendingForUpdate(sessionId);
      const maxOrder = last?.sortOrder ?? 0;
      const item: MessageQueue = {
        sessionId,
        userId,
        content,
        images,
        sortOrder: maxOrder + 1,
        status: 'PENDING',
      };
      await tx.insert(item);
      return item;
    });
  }

  /** 将消息插回队头（用于 auto-consume 失败补偿），取队首 order-1 保证排在所有现存消息之前。 */
  async enqueueHead(sessionId: number, userId: number, content: string, images: string | null): Promise<void> {
    return this.repo.transaction(async (tx) => {
      const first = await tx.findFirstPendingForUpdate(sessionId);
      const minOrder = first?.sortOrder ?? 1;
      await tx.insert({
        sessionId,
        userId,
        content,
        images,
        sortOrder: minOrder - 1,
        status: 'PENDING',
      });
    });
  }

  async dequeue(sessionId: number): Promise<MessageQueue | null> {
    const head = await this.repo.findHeadPending(sessionId);
    if (head) {
      head.status = 'DELETED';
      await this.repo.updateById(head);
    }
    return head;
  }

  async delete(queueId: number): Promise<void> {
    const item = await this.repo.findById(queueId);
    if (item) {
      item.status = 'DELETED';
      await this.repo.updateById(item);
    }
  }

  async reorder(queueId: number, direction: string): Promise<void> {
    for (let attempt = 0; ; attempt++) {
      try {
        await this.reorderOnce(queueId, direction);
        return;
      } catch (e) {
        // 相反方向并发操作同一对相邻行时锁序相反，可能触发 InnoDB 死锁；
        // 死锁由 InnoDB 即时检测并回滚一侧，基于最新状态有限重试，避免 WS 层表现为静默失败。
        // 锁等待超时不重试：等待本身已耗时 innodb_lock_wait_timeout，立即重试会进一步挂起请求。
        if (attempt < 2 && isLockConflictError(e)) continue;
        throw e;
      }
    }
  }

  private async reorderOnce(queueId: number, direction: string): Promise<void> {
    // 事务 + FOR UPDATE 锁定 current 与 neighbor（同 enqueue 的队尾锁策略），
    // 防止并发 reorder 读到相同快照导致交换丢失或重复 sort_order。
    await this.repo.transaction(async (tx) => {
      const current = await tx.findByIdForUpdate(queueId);
      if (current == null || current.status !== 'PENDING' || current.sessionId == null || current.sortOrder == null) {
        return;
      }
      const neighbor = direction === 'up'
        ? await tx.findNeighborUpForUpdate(current.sessionId, current.sortOrder)
        : await tx.findNeighborDownForUpdate(current.sessionId, current.sortOrder);
      if (neighbor != null && neighbor.sortOrder != null) {
        const tempOrder = current.sortOrder;
        current.sortOrder = neighbor.sortOrder;
        neighbor.sortOrder = tempOrder;
        await tx.updateById(current);
        await tx.updateById(neighbor);
      }
    });
  }

  listPending(sessionId: number): Promise<MessageQueue[]> {
    return this.repo.listPending(sessionId);
  }

  getById(queueId: number): Promise<MessageQueue | null> {
    return this.repo.findById(queueId);
  }

  clear(sessionId: number): Promise<void> {
    return this.repo.clearPending(sessionId);
  }
}

/** MySQL 死锁错误码：InnoDB 即时检测并自动回滚一侧，可基于最新状态立即重试。 */
function isLockConflictError(e: unknown): boolean {
  if (typeof e !== 'object' || e == null || !('code' in e)) {
    return false;
  }
  return (e as { code?: unknown }).code === 'ER_LOCK_DEADLOCK';
}
